### ISSUE TITLE
Getting the original value along with the newvalue - help needed

### ISSUE THREAD

**Issue by smatthaeus** [MrWolfZ/ngrx-forms#195]
Hi again,

[reducer.txt](https://github.com/MrWolfZ/ngrx-forms/files/4834494/reducer.txt)

Before I start with the problem, I am not sure if it is the right way to be doing things with ngrx-forms. My problem is as follows. When the user edits the form and finishes the editing (is on something like a review mode) and before submitting they want to see the initial value that was there on the form and the changed value of the form (both at the same time). 

What I tried was as follows 

in the main reducer before the formGroupReducer i check for the SetValueAction and here I THOUGHT i can get the controlid and the value. (in debug mode it does bring it up - but when compiling it does not give me the value). 

export function mainReducer(state: initialFormDataState, action: Action): FormDataState {

	let tempState;

	_if (action.type === SetValueAction.TYPE) {
		const tControl = action.controlId.split('.');
		control value = action.value;
		const control = tControl[tControl.count - 1];
		const addreducer = addControlReducer(control, value)(state.formState);
		tempState = {...tempState, formState:addReducer};
	} else { tState = state };_

	tempState = formGroupReducer(tstate.formState, action);

	switch(action.type)
	{
		case SET_FOCUS: {
			const { payload } = action as SetFocus;
			const tReducer = focusResetReducer(control)(tempState);
			return { ...tempState, formState:tempState };
		}
		default: {
			return { ...tState, formState: tState }
		}
	}

In the above code it does not compile when i say action.controlId or action.value. The idea was to invoke the addControlReducer and create a new control called prev_<controlname> and then store the values in it because I need the previous value  and the current value on the form.  The concept works fine but cannot get the control id and the value. When i hard coded it everything was good and i even saw the value in the state.

As there are so many control i would like to store the previous values only for those controls the user has changed and not every value.  The attached reducer.txt has the sample code

Would appreciate any help you can provide in this.

**Comment by MrWolfZ** [comment]
Apologies for the late response.

I hope you were able to figure this out in the meantime, but in case you still need help, here goes.

The reason the code does not compile is because TypeScript does not know the type of the action when you try to access the properties. You need to explicitly cast the action as the first action in the `if`, e.g. `const setValueAction = action  as SetValueAction`. Then you can access `setValueAction.controlId`.

Regarding storing the previous value, instead of creating a separate control for this, there is a feature called "user-defined properties" that matches this use case well. Also, and this goes for the `focusResetReducer` from your example reducer file as well, when you want to update deeply nested form controls you do not need to create the nested `updateGroup` mess, but instead you can just call the form state reducer with the action with the control ID and it will traverse the form state.

In summary, your code could look like this:

```ts
export function myStateReducer(state: MyState, action: Action) {
  let formState = state.formState;

  if (action.type === SetValueAction.TYPE) {
    const setValueAction = action as SetValueAction<any>;
    const controlId = setValueAction.controlId;
    const value = setValueAction.value;
    formState = formGroupReducer(formState, new SetUserDefinedPropertyAction(controlId, 'prevValue', value));
  }

  formState = formGroupReducer(formState, action);

  if (formState !== state.formState) {
    state = { ...state, formState };
  }

  // ... rest of your reducer
}
```

I hope this helps. If you still have questions, feel free to open a new issue.
